### PR TITLE
Assignments text

### PR DESCRIPTION
**Pull request by stellatsv**
# Description of the PR
This PR relates to #524. It contains a fix for the bug. 

In this issue, there is also mention of the existing problem that when IntelliJ is loading, the `Modules` and `Assignments` views display the wrong text. This is corrected by the time the application is ready for use and the wrongly suggested activity is safe to perform. 
### Attempts
I tried setting `InitializationActivity` as a `postStartupActity` which delayed the activity's execution a bit but it was not enough on its own. I also tried using a  [ProjectManagerListener](https://github.com/JetBrains/intellij-community/blob/master/platform/projectModel-api/src/com/intellij/openapi/project/ProjectManagerListener.java), its `projectOpened` method in particular, and from there set a boolean indicating that the project is ready, but the method got called earlier than what is needed. 

Using boolean values similar to what was used to display the correct empty text is what has helped the most so far. There is however one issue remaining with this approach when it comes to Modules view:
- When the project is an A+ project, the wrong text is visible for a little bit before displaying the tree. This is because a null `CourseViewModel` is passed to `ModulesView` during startup and there is currently no way to know if the project is an A+ project in order to display the correct text. 

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [ ] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [ ] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:133]: > When authentication is null mainViewModel is not updated?

Correct. Calling `setAuthentication` a few lines above causes `updateExercise` to be executed*. However, it's only called when the authentication read from the password storage is not null (note `ifPresent`).

---

*) Because of this line in the constructor:
```java
authentication.addValueObserver(this, MainViewModel::updateExercises);
```

**Review by OlliKiljunen — CHANGES_REQUESTED** [review]
I see this turned out to be a tricky one... Thanks for putting effort into this!

Your solution seems to work. Good! :+1: However, I'm afraid it adds quite too much complexity to solve such a "minor" thing ("minor" as its effect to the UI). And what I mean by complexity here, is, for example, that it's very difficult to reason whether those Boolean state variables in view model classes are correctly set in all cases (mutable states in general are difficult). Maintainer of the code would probably find it very hard to know if they have to do something for those manually when adding/removing some code from exercise updating logic. Concurrency brings an additional layer of uncertainty.

The reason why this turned out to be so tricky may be that the procedure that handles creation of main view model etc. was/is poorly designed and even flawed. @nikke234's open PR #560 improves and, especially, _simplifies_ that a lot, which may help it solve this issue too.

So what I suggest is that you make those minor changes I requested in my inline comments (also to make this PR easier to understand for further reviewers), then we wait for #560 to be merged (should be soon) and see together how things look after that (let's add @nikke234 as a reviewer then, too).

You've put a great effort in this task and that is, in any case, highly valuable for the project. Hopefully, you're not already totally exhausted of fighting this bug (and if you are, it's ok if you want to work with other tickets before returning to this).

By the way, I couldn't help noticing a not-too-big refactoring need closely related to this PR. It's when that `emptyModel` is created in `MainViewModel::updateExercises`. It would be more OO design, to have `EmptyExercisesTreeViewModel` as a separate subclass of `ExercisesTreeViewModel`. Then `isEmptyTextVisible` field wouldn't be required because the base class would always return true for `isEmptyTextVisible()` and the subclass would override it to return always false. Would you like to make that change (as a part of this PR, or separately)? Not necessary, just an idea.

File: src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/ExercisesTreeViewModel.java
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/ExercisesTreeViewModel.java:18]: Hence, all calls to `ExercisesTreeViewModel::setProjectReady` should be removed as they have no effect anywhere. Please correct me if I'm wrong.

Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/ExercisesTreeViewModel.java:16]: The getter is used in `ExercisesView`.

File: src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/EmptyExercisesTreeViewModel.java
Comment (nikke234) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/EmptyExercisesTreeViewModel.java:17]: Since it's supposed to be empty, I would suggest that this constructor doesn't take a `List` parameter at all, and instead it always passes an empty list to the superclass constructor

File: src/main/resources/META-INF/plugin.xml
Comment (nikke234) [inline src/main/resources/META-INF/plugin.xml:24]: Do you have a link or something where I could read about the differences of these two? My googling isn't bringing results :smile:

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (nikke234) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:77]: I think the inner `if (course == null)` could be removed and just directly assign `new EmptyExercise...` to `emptyModel`

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (nikke234) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:90]: `auth` is never `null` at this point, because the earlier `if (course == null || auth == null)` ends with a return.

**Review by nikke234 — COMMENTED** [review]
Nice! I left a few comments about code clarity. It's a bit tricky to reason and be confident that all the booleans get set correctly, so I'll double-check that and see if the newest merges could simplify anything here

File: src/main/resources/META-INF/plugin.xml
Comment (stellatsv) [reply src/main/resources/META-INF/plugin.xml:24]: https://plugins.jetbrains.com/docs/intellij/plugin-components.html#project-open

Comment (nikke234) [reply src/main/resources/META-INF/plugin.xml:24]: Thanks!

Running that initialization activity on the EDT (more familiarly the "UI thread") should be avoided, because the UI freezes until the activity is done. This is especially important in this case, because the initialization activity fetches network resources, which can have long and unpredictable delays. While running the code I occasionally got log entries like this:
```
2021-03-14 23:04:32,865 [  20349]   WARN - .diagnostic.PerformanceWatcher - UI was frozen for 12581ms, details saved to /l/drosden1/intellij-plugin/build/idea-sandbox/system/log/threadDumps-freeze-20210314-230425-IC-203.7148.57-ObservableReadWriteProperty.get-12sec
```
Then again running it in the background makes the rest more difficult. The more I work on this issue, the more I realize how complicated it is to solve it :smile: I'm personally at least tempted to reconsider whether this should be "release-blocking", let's discuss in tomorrow's daily.

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (nikke234) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:76]: This doesn't have to be fixed now, but it's quite confusing that `new ExercisesTreeViewModel(new ArrayList<>(), new Options)` is actually different from `new EmptyExercisesTreeViewModel()`, because the first uses the empty text of the exercises tree, whereas the second one hides the tree and instead shows the separate empty text label. One option would be renaming `isEmptyTextVisible` to something else, but then `EmptyExercisesTreeViewModel` should also be renamed. Anyways, doesn't need to be fixed now

File: src/main/resources/resources.properties
Comment (nikke234) [inline src/main/resources/resources.properties:148]: Nit: I would use three dots :smile:

**Review by nikke234 — APPROVED** [review]
Looks good :+1: Definitely not the easiest issue to solve :grimacing:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/ProjectKey.java
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/ProjectKey.java:11]: ❤️

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:132]: I'm not sure if it makes sense to ensure `updateExercises()` is called here. That should perhaps be done by the caller.

File: src/test/java/fi/aalto/cs/apluscourses/intellij/utils/ProjectKeyTest.java
Comment (OlliKiljunen) [inline src/test/java/fi/aalto/cs/apluscourses/intellij/utils/ProjectKeyTest.java:10]: 😍:pray:

**Review by OlliKiljunen — APPROVED** [review]
Great work! And sorry for a late review.

Thanks for fighting this issue and congrats for beating it for now. 😄

File: src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:132]: BTW what's the reason that the call is needed here in the first place?

Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java:132]: There is an earlier comment in this PR that sheds some light. I am not 100% sure about this but I think the view was not updated in a case where there was no token set and I needed to manually call updateExercises. I tested that everything works without it now so I can remove it.

**Pull request merged**
